From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 22 Jul 2023 21:48:49 +0200
Subject: [PATCH] Client perspective - Item replacements - Pass context of
 items in item frames

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
index debb34a8f99c5b1356a3f742ae34a002c375f1e8..7cc4a5805b1e4f2df555b8ed4280123fea86e7c0 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundSetEntityDataPacket.java
@@ -10,11 +10,15 @@ import net.minecraft.network.syncher.SynchedEntityData;
 public class ClientboundSetEntityDataPacket extends org.fiddlemc.fiddle.packet.PacketWithExplicitTargetClientConnection<ClientGamePacketListener> {
     private final int id;
     private final List<SynchedEntityData.DataValue<?>> packedItems;
+    // Fiddle start - client perspective - item replacements - pass context of items in item frames
+    private final @org.jetbrains.annotations.Nullable org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext itemStackContext;
 
-    public ClientboundSetEntityDataPacket(int id, List<SynchedEntityData.DataValue<?>> packedItems) {
+    public ClientboundSetEntityDataPacket(int id, List<SynchedEntityData.DataValue<?>> packedItems, @org.jetbrains.annotations.NotNull org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext itemStackContext) {
+        // Fiddle end - client perspective - item replacements - pass context of items in item frames
         super();
         this.id = id;
         this.packedItems = packedItems;
+        this.itemStackContext = itemStackContext; // Fiddle - client perspective - item replacements - pass context of items in item frames
     }
 
     public int id() {
@@ -29,12 +33,12 @@ public class ClientboundSetEntityDataPacket extends org.fiddlemc.fiddle.packet.P
     public static final int EOF_MARKER = 255;
 
     public ClientboundSetEntityDataPacket(FriendlyByteBuf buf) {
-        this(buf.readVarInt(), unpack(buf));
+        this(buf.readVarInt(), unpack(buf), null); // Fiddle - client perspective - item replacements - pass context of items in item frames
     }
 
-    private static void pack(List<SynchedEntityData.DataValue<?>> trackedValues, FriendlyByteBuf buf) {
+    private static void pack(List<SynchedEntityData.DataValue<?>> trackedValues, FriendlyByteBuf buf, @org.jetbrains.annotations.Nullable org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext itemStackContext) { // Fiddle - client perspective - item replacements - pass context of items in item frames
         for(SynchedEntityData.DataValue<?> dataValue : trackedValues) {
-            dataValue.write(buf);
+            dataValue.write(buf, itemStackContext); // Fiddle - client perspective - item replacements - pass context of items in item frames
         }
 
         buf.writeByte(255);
@@ -54,7 +58,7 @@ public class ClientboundSetEntityDataPacket extends org.fiddlemc.fiddle.packet.P
     @Override
     public void write(FriendlyByteBuf buf) {
         buf.writeVarInt(this.id);
-        pack(this.packedItems, buf);
+        pack(this.packedItems, buf, this.itemStackContext); // Fiddle - client perspective - item replacements - pass context of items in item frames
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/network/syncher/EntityDataSerializers.java b/src/main/java/net/minecraft/network/syncher/EntityDataSerializers.java
index 9ca897d92c5bdd2764d114c74d64c776674d6beb..10b6947ccda5abc4c696e9ee6fdc08964ff99b23 100644
--- a/src/main/java/net/minecraft/network/syncher/EntityDataSerializers.java
+++ b/src/main/java/net/minecraft/network/syncher/EntityDataSerializers.java
@@ -39,10 +39,20 @@ public class EntityDataSerializers {
     public static final EntityDataSerializer<String> STRING = EntityDataSerializer.simple(FriendlyByteBuf::writeUtf, FriendlyByteBuf::readUtf);
     public static final EntityDataSerializer<Component> COMPONENT = EntityDataSerializer.simple(FriendlyByteBuf::writeComponent, FriendlyByteBuf::readComponent);
     public static final EntityDataSerializer<Optional<Component>> OPTIONAL_COMPONENT = EntityDataSerializer.optional(FriendlyByteBuf::writeComponent, FriendlyByteBuf::readComponent);
-    public static final EntityDataSerializer<ItemStack> ITEM_STACK = new EntityDataSerializer<ItemStack>() {
+    // Fiddle start - client perspective - item replacements - pass context of items in item frames
+    public static abstract class ItemStackEntityDataSerializer implements EntityDataSerializer<ItemStack> {
+
+        public void write(FriendlyByteBuf buf, ItemStack value, @org.jetbrains.annotations.NotNull org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext itemStackInPacketContext) {
+            var sanitizedItemStack = net.minecraft.world.entity.LivingEntity.sanitizeItemStack(value, true);
+            buf.writeItem(sanitizedItemStack, itemStackInPacketContext); // Paper start - prevent oversized data
+        }
+
+    }
+    public static final EntityDataSerializer<ItemStack> ITEM_STACK = new ItemStackEntityDataSerializer() {
+        // Fiddle end - client perspective - item replacements - pass context of items in item frames
         @Override
         public void write(FriendlyByteBuf buf, ItemStack value) {
-            buf.writeItem(net.minecraft.world.entity.LivingEntity.sanitizeItemStack(value, true)); // Paper - prevent oversized data
+            throw new UnsupportedOperationException(); // Fiddle - client perspective - item replacements - pass context of items in item frames
         }
 
         @Override
diff --git a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
index 5dfb35117c285e0b202dc9c088ad5848beb8d054..b8603e5424449f34d443d8acdb4229e8d775a720 100644
--- a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
@@ -279,7 +279,7 @@ public class SynchedEntityData {
 
             if (list != null) {
                 if (to.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper
-                to.connection.send(new ClientboundSetEntityDataPacket(this.entity.getId(), list));
+                to.connection.send(new ClientboundSetEntityDataPacket(this.entity.getId(), list, this.entity.getItemStackInPacketContext())); // Fiddle - client perspective - item replacements - pass context of items in item frames
                 } // Paper
             }
         }
@@ -326,7 +326,7 @@ public class SynchedEntityData {
             values.add(synchedValue.value());
         }
 
-        to.connection.send(new ClientboundSetEntityDataPacket(this.entity.getId(), values));
+        to.connection.send(new ClientboundSetEntityDataPacket(this.entity.getId(), values, this.entity.getItemStackInPacketContext())); // Fiddle - client perspective - item replacements - pass context of items in item frames
     }
     // Paper end
 
@@ -380,7 +380,7 @@ public class SynchedEntityData {
             return new SynchedEntityData.DataValue<>(data.getId(), datawatcherserializer, datawatcherserializer.copy(value));
         }
 
-        public void write(FriendlyByteBuf buf) {
+        public void write(FriendlyByteBuf buf, @Nullable org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext itemStackInPacketContext) { // Fiddle - client perspective - item replacements - pass context of items in item frames
             int i = EntityDataSerializers.getSerializedId(this.serializer);
 
             if (i < 0) {
@@ -388,7 +388,13 @@ public class SynchedEntityData {
             } else {
                 buf.writeByte(this.id);
                 buf.writeVarInt(i);
-                this.serializer.write(buf, this.value);
+                // Fiddle start - client perspective - item replacements - pass context of items in item frames
+                if (this.serializer instanceof EntityDataSerializers.ItemStackEntityDataSerializer itemStackSerializer) {
+                    itemStackSerializer.write(buf, (net.minecraft.world.item.ItemStack) this.value, Objects.requireNonNull(itemStackInPacketContext));
+                } else {
+                    this.serializer.write(buf, this.value);
+                }
+                // Fiddle end - client perspective - item replacements - pass context of items in item frames
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 81d0b2933040a451441f660f9e46199ae3b111e3..a604fb849c5c880233b416f95c57ebb060a3fe2e 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -300,7 +300,7 @@ public class ServerEntity {
         this.yHeadRotp = Mth.floor(this.entity.getYHeadRot() * 256.0F / 360.0F);
         sender.accept(packet);
         if (this.trackedDataValues != null) {
-            sender.accept(new ClientboundSetEntityDataPacket(this.entity.getId(), this.trackedDataValues));
+            sender.accept(new ClientboundSetEntityDataPacket(this.entity.getId(), this.trackedDataValues, this.entity.getItemStackInPacketContext())); // Fiddle - client perspective - item replacements - pass context of items in item frames
         }
 
         boolean flag = this.trackDelta;
@@ -375,7 +375,7 @@ public class ServerEntity {
 
         if (list != null) {
             this.trackedDataValues = datawatcher.getNonDefaultValues();
-            this.broadcastAndSend(new ClientboundSetEntityDataPacket(this.entity.getId(), list));
+            this.broadcastAndSend(new ClientboundSetEntityDataPacket(this.entity.getId(), list, this.entity.getItemStackInPacketContext())); // Fiddle - client perspective - item replacements - pass context of items in item frames
         }
 
         if (this.entity instanceof LivingEntity) {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 305b43071aa1cf8feee75fae757bb7734ae33771..8eab98754d520d74c3f960fa2170bd510f87b23c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4789,4 +4789,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return ((net.minecraft.server.level.ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
     }
     // Paper end
+
+    // Fiddle start - client perspective - item replacements - pass context of items in item frames
+    public @org.jetbrains.annotations.NotNull org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext getItemStackInPacketContext() {
+        return org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext.DEFAULT;
+    }
+    // Fiddle end - client perspective - item replacements - pass context of items in item frames
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
index 955316687e2e29ad75a0052317a7b0f89034c82a..29ca6d6993cd745647706e20b9fbf0eacff64531 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -539,4 +539,12 @@ public class ItemFrame extends HangingEntity {
 
         return (float) Mth.wrapDegrees(180 + enumdirection.get2DDataValue() * 90 + this.getRotation() * 45 + i);
     }
+
+    // Fiddle start - client perspective - item replacements - pass context of items in item frames
+    @Override
+    public @org.jetbrains.annotations.NotNull org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext getItemStackInPacketContext() {
+        return org.fiddlemc.fiddle.packet.item.ItemStackInPacketContext.IN_ITEM_FRAME;
+    }
+    // Fiddle end - client perspective - item replacements - pass context of items in item frames
+
 }
